torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -260,8 +260,134 @@ void BatchMMBlock(Block* block) {
+    // a lot of independent MMs, that depend on the first one, but I doubt this will
+    // be a common scenario.
+    for (size_t i = 0; i < mms.size(); ++i) {
+      if (mms[i] == nullptr) continue;
+      for (size_t j = i + 1; j < mms.size(); ++j) {
+        if (mms[j] == nullptr) continue;
+        if (!mms[j]->couldMoveBeforeTopologically(mms[i])) {
+          mms[j] = nullptr;
+        }
+      }
+    }
+    return filter(mms, [](Node *n) { return n != nullptr; });
+  };
+
+  Block * block = value->node()->owningBlock();
+  std::vector<Node*> lhses; // Will contain nodes where value is used as an lhs
+  std::vector<Node*> rhses; // Like above, but rhs
+  for (Use u : value->uses()) {
+    if (u.user->owningBlock() == block &&
+        u.user->matches("aten::mm(Tensor self, Tensor mat2) -> Tensor")) {
+      if (u.offset == 0 && u.user->inputs()[1] != value) {
+        lhses.push_back(u.user);
+      } else if (u.offset == 1 && u.user->inputs()[0] != value) {
+        rhses.push_back(u.user);
+      }
+    }
+  }
+  return std::make_pair(postprocess(lhses), postprocess(rhses));
+}
+
+void BatchMMSide(Block * block) {
+  // NB: 8 is the current loop unrolling factor
+  static constexpr size_t how_many_is_many = 8;
+  static const auto batch_side = [](std::vector<Node*>& mms, Side side) {
+    JIT_ASSERT(!mms.empty());
+    WithInsertPoint insert_guard { mms[0] };
+    Graph* graph = mms[0]->owningGraph();
+    Node* batch_mm = graph->create(Symbol::prim("MMBatchSide"),
+                                    /*inputs=*/{}, /*num_outputs=*/mms.size());
+    graph->insertNode(batch_mm);
+    batch_mm->i_(Symbol::attr("side"), static_cast<int>(side));
+    Value* const_side = mms[0]->inputs().at(side == Side::LHS ? 0 : 1);
+    batch_mm->addInput(const_side);
+    for (size_t i = 0; i < mms.size(); ++i) {
+      batch_mm->addInput(mms[i]->inputs().at(side == Side::LHS ? 1 : 0));
+      mms[i]->output()->replaceAllUsesWith(batch_mm->outputs().at(i));
+    }
+  };
+
+  std::unordered_set<Value*> considered_values;
+  for (Node * node : block->nodes()) {
+    if (node->matches("aten::mm(Tensor self, Tensor mat2) -> Tensor")) {
+      for (Value * input : node->inputs()) {
+        if (/*bool not_inserted = */!considered_values.emplace(input).second) {
+          continue;
+        }
+        auto uses_with_many = gatherIndependentMMUses(input);
+        if (uses_with_many.first.size() >= how_many_is_many) {
+          batch_side(uses_with_many.first, Side::LHS);
+        }
+        if (uses_with_many.second.size() >= how_many_is_many) {
+          batch_side(uses_with_many.second, Side::RHS);
+        }
+      }
+    } else {
+      for (Block* subblock : node->blocks()) {
+        BatchMMSide(subblock);
+      }
+    }
+  }
+
+}
+
 void BatchMM(std::shared_ptr<Graph>& graph) {
-  BatchMMBlock(graph->block());
+  BatchMMTreeReduce(graph->block());
+  BatchMMSide(graph->block());
```

Comment:
It might be worth adding a test for BatchMMSide (expect test?) so that we will notice if the pass stops working